First draft of embedding management (#283)
vespa-engine/blog: _posts/2023-03-31-tailoring-frozen-embeddings-with-vespa.md

for the home page. We also hear they are discussing a related
articles feature, where for each article, the site can suggest related articles. 
At the end of the year, we will face the challenge of operationalizing three different
embedding-based use cases. There must be a better way?

## Frozen Embeddings to the Rescue
![Frozen embeddings](/assets/2023-03-31-tailoring-frozen-embeddings-with-vespa/image6.png)
An emerging industry trend that addresses operationalizing related
complexity is to use [frozen foundational
embeddings](https://ai.facebook.com/blog/multiray-large-scale-AI-models/)
that can be [reused
](https://medium.com/pinterest-engineering/searchsage-learning-search-query-representations-at-pinterest-654f2bb887fc) for
different tasks without incurring the usual costs related to embedding
versioning, storage, or inference infrastructure.

![Vector Space Illustration](/assets/2023-03-31-tailoring-frozen-embeddings-with-vespa/image2.png)

_A frozen model embeds data Q and D into vector space. By transforming
the Q representation to Q’, the vector distance is reduced

> `(d(D,Q) < d(D,Q’)`. This figure illustrates fine-tuning of metric

Reviewer: Which distance is smallest? `d(D,Q)` or `d(D,Q’)`? Based on the description, it sounded like it should be `d(D,Q) > d(D,Q’)` instead.

Author: Thanks!

distances over frozen embeddings. Note that the D representation
does not change, which is a practical property for search and
recommendation use cases with potentially billions of embedding
representations of items._

With frozen embeddings, the data is embedded once using a foundational
embedding model. Developers can then adapt the representation to
specific tasks by adding transformation layers on the frozen
embeddings. The following sections describe three different methods
for adapting frozen embeddings for search or recommendation use
cases and how these approaches are represented in Vespa.

> * Tuning the query tower in two-tower embedding models 

Reviewer: There is something strange with these 3 sentences. Leftovers?

Reviewer: Sorry, I didn't see this was a list. Ignore the comment.

* Simple query embedding transformations 
* Advanced transformations using Deep Neural Networks

## Two-tower models

The industry standard for semantic vector search is using a two-tower
architecture based on a [Transformer](https://en.wikipedia.org/wiki/Transformer_(machine_learning_model)) based model. 

This architecture is also called a _bi-encoder_ model, as there is a query and document
encoder.  Most of the two-tower architecture models use the same
weights for both the query and document encoder. This is not ideal
as if we tune the model, we would need to re-embed all our items
again. By de-coupling the model weights of the query and document
tower, developers can treat the document embeddings as frozen,
to avoid re-encoding all the items. When fine-tuning the model for the
specific task, developers can instead just tune the query tower and
leave the frozen document tower alone.

![Frozen Query Tower](/assets/2023-03-31-tailoring-frozen-embeddings-with-vespa/image3.png)

The frozen document tower and embeddings significantly reduce the

> complexity and cost of serving and training. For training the query

Reviewer: Long and complex sentence starting with "For training the query...". Consider rewrite.

tower, because there is no need to encode the document vector over
again, the document’s embedding representation can be fetched
directly from Vespa during training. This saves at least 2x of
computational complexity during training as one does not need to

> run inference with the document tower during training. In practice,

Reviewer: ```suggestion
run inference with the document tower. In practice,
```

since documents are longer than queries and Transformer models
scales quadratic with input lengths, the computational saving is
higher than 2x.

On the serving side in Vespa, there is no need to re-process the

> frozen document embedding, this saves the compute of performing the

Reviewer: ```suggestion
frozen document embedding. This saves the compute of performing the
```

}
</pre>

The learned weights are exported from any ML framework (e.g.,
[PyTorch](https://pytorch.org/),
[scikit-learn](https://scikit-learn.org/stable/)) used to train the
matrix weights. And the weights are exported to a [constant
tensor](https://docs.vespa.ai/en/tensor-user-guide.html#constant-tensors)
file. Meanwhile, the `transform_query` function [performs a vector
matrix
product](https://docs.vespa.ai/playground/#N4KABGBEBmkFxgNrgmUrWQPYAd5QGNIAaFDSPBdDTAF30gDUBTA2rAJwFoBbAQ1ocAlgA8wODlgAmAVzYA6MADEhHAM60wAd2ZgpzaEIB2uvmABurdhzAADRrbB8jUp2H6DRdgOq24kMlQAX0Cg0gxqclwGZhJAiAp8SJpIIwZGOJpMWIRIWmYjNU4AChFEAGYAXQBKOEQABnkAVmIwRpa25sqArJCMMMDkhOjc2PCUyjR4zDTc70ysqByofMKSrUQAFkriMqraxEQARnl61pOj8-kAJh2ka9PWh8uwB9viY8ewC6v3z7PvvIXidbt1pn1guMIEMoCNCAtyJMYQl6LkAIIWKycXgCYRiCTSOSaIRqMC0AAWuh4MgANrQhDgaUICAIhFgjGAsNAyZTMWxOE4XO5cV5nK41DIeDxjABzTmWGwU3QELBS9l6IQ8ApqNlGfzg0JQqZZbCTSBjaaJKjTBKzKAIlLLSASnjFRhgABUYG8rS01R6NAhEAhlRAQSAA),
returning a modified vector of the same dimensionality.

This representation is used to score the documents in the first-phase
ranking expression. Note that this is effectively represented as
[a re-ranking phase](https://docs.vespa.ai/en/phased-ranking.html)

> as the query tensor used for the nearestNeighbor search is untouched.

Reviewer: ```suggestion
as the query tensor used for the [nearestNeighbor](https://docs.vespa.ai/en/reference/query-language-reference.html#nearestneighbor) search is untouched.
```

The weight tensor does not necessarily need to be a constant across
all users. For example, one can have a weight tensor per user, as
shown in the [recommendation use
case](https://docs.vespa.ai/en/tutorials/news-4-embeddings.html),
to unlock true personalization.

## Advanced transformation using Deep Neural Networks

Another approach for customization is to use the query and document
embeddings as input to another Deep Neural Network (DNN) model.
This approach can be combined with the previously mentioned
approaches because it's applied as a re-scoring model in a [phased

> ranking ](https://docs.vespa.ai/en/phased-ranking.html)pipeline.

Reviewer: ```suggestion
ranking](https://docs.vespa.ai/en/phased-ranking.html)pipeline.
```

		super(CustomEmbeddingSimilarity, self).__init__()
		self.fc1 = nn.Linear(2*dimensionality, 256)
		self.fc2 = nn.Linear(256, 128)
		self.fc3 = nn.Linear(128, 64)
		self.fc4 = nn.Linear(64, 1)
	def forward(self, query , document):
		x = torch.cat((query, document), dim=1)
		x = nn.functional.relu(self.fc1(x))
		x = nn.functional.relu(self.fc2(x))
		x = nn.functional.relu(self.fc3(x))
		return torch.sigmoid(self.fc4(x))
dim = 384
ranker = CustomEmbeddingSimilarity(dimensionality=dim)
# Train the ranker model ..
# Export to ONNX for inference with Vespa 
input_names = ["query","document"]
output_names = ["similarity"]
document = torch.ones(1,dim,dtype=torch.float)
query = torch.ones(1,dim,dtype=torch.float)
args = (query,document)
torch.onnx.export(ranker,
                  args=args,
                  f="custom_similarity.onnx",
                  input_names = input_names,
                  output_names = output_names,
                  opset_version=15)

</pre>
The above [PyTorch](https://pytorch.org/) [model.py](https://github.com/vespa-engine/sample-apps/blob/master/custom-embeddings/model.py) snippet defines a custom
DNN-based similarity model which takes the query and document

> embedding as input. This model is exported to[ ONNX](https://onnx.ai/)

Reviewer: ```suggestion
embedding as input. This model is exported to [ONNX](https://onnx.ai/)
```

format for [accelerated
inference](https://blog.vespa.ai/stateful-model-serving-how-we-accelerate-inference-using-onnx-runtime/)
using Vespa’s support for [ranking with
ONNX](https://docs.vespa.ai/en/onnx.html) models.

> Notice that [PyTorch ](https://pytorch.org/)model uses batch

Reviewer: ```suggestion
Notice that the [PyTorch](https://pytorch.org/) model uses batch
```

 We still need to serve the current embedding model which is in production, used to embed new documents and the current real-time stream of queries. 
 
* Index the new vector embedding representation to the
serving infrastructure that we use for efficient vector search. If
we are fortunate enough to be using [Vespa](https://vespa.ai/), which supports 
[nearest neighbor search with filtering](https://blog.vespa.ai/constrained-approximate-nearest-neighbor-search/),
we can index the new embedding in a new field, without duplicating
other schema fields. Adding the new vector field still adds to the serving cost,
as we double the resource usage footprint related to indexing and storage. 

* After all this, we are finally ready to evaluate the new embedding
model online. Depending on the outcome of the online evaluation, we can
garbage collect either the new or old embedding data. 

That's a lot of complexity and cost to evaluate a model online, but now we can relax? 
Wait, our PM now wants to introduce [news article recommendations](https://docs.vespa.ai/en/tutorials/news-4-embeddings.html)

> for the home page, and the ML team is planning on using embeddins for this project. We also hear they are 

Reviewer: ```suggestion
for the home page, and the ML team is planning on using embeddings for this project. We also hear they are 
```

documents, as the same input will produce the exact same frozen document embedding representation. 
This saves the compute of performing the
inference and avoids introducing embedding versioning. And,
because Vespa allows deploying multiple query tower models, 
applications may test the accuracy of new models, without re-processing documents, 
which allows for frequent model deployment and evaluations.

Managing complex infrastructure for producing text embedding vectors
could be challenging, especially at query serving time, with low
latency, high availability, and high query throughput. Vespa allows
developers to [represent embedding
models](https://blog.vespa.ai/text-embedding-made-simple/) in Vespa.
Consider the following schema, expressed using [Vespa’s schema
definition language](https://docs.vespa.ai/en/schemas.html):

<pre>
schema doc {
  document doc {
    field text type string {}  
  } 
  field embedding type tensor&lt;float&gt;(x[384]) { 
    indexing: input text | embed frozen | attribute | index
  }
}
</pre>

In this case, Vespa will produce embeddings using a `frozen` embedding
model, and at query time, we can either use the frozen model to

> encode the query or a new fine-tuned model. Deplying multipe query tower models allows

Reviewer: ```suggestion
encode the query or a new fine-tuned model. Deploying multiple query tower models allows
```
